### PR TITLE
Compare type by (name,type) pair rather than (index,type) pair during Parquet's schema mismatch checking

### PR DESCRIPTION
**Pull request by beinan**
Regarding parquet's schema mismatch checking, the existing implementation is selecting each sub-field by its index, which might cause fake alert of mismatching. e.g.
struct<a bigint, b string>. might mismatch parquet's group type {b binary, a int64}, but actually they are the same type.

So we add a map of <name, type> to compare each sub-field by name->type pair rather than index->type pair.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 1]
```diff
@@ -322,8 +325,19 @@ private static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTy
             switch (prestoType) {
                 case ROW:
                     if (groupType.getFields().size() == type.getTypeParameters().size()) {
+                        checkState(type instanceof RowType, "It must be a RowType here.");
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: use checkArgument

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 2]
```diff
@@ -322,8 +325,19 @@ private static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTy
...
+                        RowType rowType = (RowType) type;
+                        Map<String, Type> prestoFieldMap = rowType.getFields().stream().collect(
+                                Collectors.toMap(
+                                        f -> f.getName().get(),
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: s/f/field/g

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 3]
```diff
@@ -322,8 +325,19 @@ private static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTy
...
+                                        f -> f.getType()));
                         for (int i = 0; i < groupType.getFields().size(); i++) {
-                            if (!checkSchemaMatch(groupType.getFields().get(i), type.getTypeParameters().get(i))) {
+                            org.apache.parquet.schema.Type parquetFieldType = groupType.getFields().get(i);
+                            Type prestoFieldType = prestoFieldMap.get(parquetFieldType.getName());
+                            if (prestoFieldType == null) {
+                                prestoFieldType = prestoFieldMap.get(parquetFieldType.getName() + "_");
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: why use parquetFieldType.getName() + "_"?
if prestoFieldType == null, we should return false?

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 1 above)
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 2 above)
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 3 above)
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: Good call, this feature is twitter specific, which shouldn't in this pr

**Comment by beinan** [comment]
Updates:
The comparison on sub-field names should be case insensitive.
Add a couple of test cases.

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 4]
```diff
@@ -322,8 +325,17 @@ private static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTy
             switch (prestoType) {
                 case ROW:
                     if (groupType.getFields().size() == type.getTypeParameters().size()) {
+                        checkArgument(type instanceof RowType, "It must be a RowType here.");
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: do we need to check type here? prestoType is from type, which is ROW. I think we do not need this line

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 5]
```diff
@@ -722,6 +723,90 @@ public void testSchemaMismatch()
                 .isFailingForPageSource(new ParquetPageSourceFactory(TYPE_MANAGER, HDFS_ENVIRONMENT, STATS, new HadoopFileOpener()), expectedErrorCode, expectedMessageRowLongNest);
     }
 
+    @Test
+    public void testSchemaMismatchOnNestedStruct()
+            throws Exception
+    {
+        //testing fields order does NOT matter
```
Comment (zhenxiao) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: test out of order fields in nested Row type

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 6]
```diff
@@ -722,6 +723,90 @@ public void testSchemaMismatch()
...
+                mapBlockOf(createUnboundedVarcharType(), new ArrayType(RowType.anonymous(ImmutableList.of(DOUBLE, INTEGER))),
+                        "test", arrayBlockOf(RowType.anonymous(ImmutableList.of(DOUBLE, INTEGER)), rowBlockOf(ImmutableList.of(DOUBLE, INTEGER), 5.0, 1L))));
+        assertThatFileFormat(PARQUET)
+                .withWriteColumns(ImmutableList.of(writeColumn))
+                .withReadColumns(ImmutableList.of(readColumn))
+                .withRowsCount(1)
+                .withSession(parquetPageSourceSession)
+                .isReadableByPageSource(new ParquetPageSourceFactory(TYPE_MANAGER, HDFS_ENVIRONMENT, STATS, new HadoopFileOpener()));
+
+        //testing field name case insensitive.
```
Comment (zhenxiao) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: test field name case sensitivity in nested Row type

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java [hunk 7]
```diff
@@ -722,6 +723,90 @@ public void testSchemaMismatch()
...
+                mapBlockOf(createUnboundedVarcharType(), new ArrayType(RowType.anonymous(ImmutableList.of(DOUBLE, INTEGER))),
+                        "test", arrayBlockOf(RowType.anonymous(ImmutableList.of(DOUBLE, INTEGER)), rowBlockOf(ImmutableList.of(DOUBLE, INTEGER), 5.0, 1L))));
+        assertThatFileFormat(PARQUET)
+                .withWriteColumns(ImmutableList.of(writeColumn))
+                .withReadColumns(ImmutableList.of(readColumn))
+                .withRowsCount(1)
+                .withSession(parquetPageSourceSession)
+                .isReadableByPageSource(new ParquetPageSourceFactory(TYPE_MANAGER, HDFS_ENVIRONMENT, STATS, new HadoopFileOpener()));
+
+        //testing rename a sub-field would cause a failure.
```
Comment (zhenxiao) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: test field name mismatch in nested Row type

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java [hunk 8]
```diff
@@ -322,8 +325,17 @@ private static boolean checkSchemaMatch(org.apache.parquet.schema.Type parquetTy
...
+                        Map<String, Type> prestoFieldMap = rowType.getFields().stream().collect(
+                                Collectors.toMap(
+                                        field -> field.getName().get().toLowerCase(Locale.ENGLISH),
+                                        field -> field.getType()));
                         for (int i = 0; i < groupType.getFields().size(); i++) {
-                            if (!checkSchemaMatch(groupType.getFields().get(i), type.getTypeParameters().get(i))) {
+                            org.apache.parquet.schema.Type parquetFieldType = groupType.getFields().get(i);
+                            String fieldName = parquetFieldType.getName().toLowerCase(Locale.ENGLISH);
+                            Type prestoFieldType = prestoFieldMap.get(fieldName);
+                            if (prestoFieldType == null || !checkSchemaMatch(parquetFieldType, prestoFieldType)) {
```
Comment (zhenxiao) [inline presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: do we need to error out when prestoFieldType == null? If parquet file has a field not exist in Presto metastore, it is OK we just leave it there. What do you think?

**Comment by zhenxiao** [comment]
BTW, could you please change the first commit message to "use field name for Parquet schema mismatch checking"

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 4 above)
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: good call, removed

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java (hunk 5 above)
Comment (beinan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java (hunk 6 above)
Comment (beinan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: done

File: presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java (hunk 7 above)
Comment (beinan) [reply presto-hive/src/test/java/com/facebook/presto/hive/TestHiveFileFormats.java]: done

File: presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java (hunk 8 above)
Comment (beinan) [reply presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetPageSourceFactory.java]: change the condition to prestoFieldType != null && !checkSchemaMatch(parquetFieldType, prestoFieldType)

**Comment by beinan** [comment]
> BTW, could you please change the first commit message to "use field name for Parquet schema mismatch checking"

changed

**Comment by zhenxiao** [comment]
thank you @beinan 
looks good to me. I will merge upon tests completes

**Comment by zhenxiao** [comment]
seems having build error. @beinan could you please rebase to the master?

**Pull request merged**